eclass/rocm.eclass

```diff
@@ -15,9 +15,13 @@
 # edit USE flag to control which GPU architecture to compile. Using
 # ${ROCM_USEDEP} can ensure coherence among dependencies. Ebuilds can call the
 # function get_amdgpu_flag to translate activated target to GPU compile flags,
-# passing it to configuration. Function check_amdgpu can help ebuild ensure
+# passing it to configuration. Function rocm_use_hipcc switches active compiler
+# to hipcc and cleans incompatible flags (useful for users with gcc-only flags
+# in /etc/portage/make.conf). Function check_amdgpu can help ebuild ensure
 # read and write permissions to GPU device in src_test phase, throwing friendly
-# error message if unavailable.
+# error message if unavailable. However src_configure in general should not
+# access any GPU devices. If it does, it usually means that CMakeLists.txt
```

Comment:
BTW, have you observed any package that probes GPU device in `src_configure`?

Reply:
Anything CUDA related if you don't specify an architecture.

Reply:
OK, that sounds reasonable. And what should we do for those packages? Add `addpredict` in the `src_configure` of their ebuilds?

Reply:
I'm very much against addpredict. You end up with silent failures that are very hard to detect.

Either you fail hard and completely avoid the detection. Or we actually allow the detection of the native device.

Reply:
I rechecked all issues that caused sandbox violations in the past for ROCm libraries.

The main root cause was hipcc: it automatically discovers gpu just when user calls `hipcc test.cpp` - see https://wiki.gentoo.org/wiki/HIP#hipcc_.28Clang_wrapper.29 . This call happens every time when cmake tests that C/C++ "is able to compile a simple test program". However, many ebuilds use hipcc without addpredict now and almost nobody sees sandbox violation errors is because of improvements in amd_rocm_enumerator implementation. 

As described in https://github.com/ROCm/rocminfo?tab=readme-ov-file#execution, enumeration consists of 4 steps:
1) read file from ROCM_TARGET_LST env variable (if any)
2) call `cat /sys/class/kfd/kfd/topology/nodes/*/properties | grep gfx`
3) call `/usr/bin/lspci -n -d 1002:` and try to match with a dictionary
4) finally, if nothing was found during 1-3, call rocminfo (which causes sandbox violation, if kfd device exists)

After all recent changes, autodetection still happens, but it does not cause sandbox errors because execution never reaches step 4. However we can do better and effectively disable 2 and 3 with `ROCM_TARGET_LST` file. I updated `rocm_use_hipcc` adding the following:

```bash
	local target_lst="${BUILD_DIR}"/gentoo_rocm_target.lst
	if [[ "${AMDGPU_TARGETS[@]}" = "" ]]; then
		# Expected no GPU code; still need to calm down sandbox
		echo "gfx000" > "${target_lst}"
	else
		printf "%s\n" ${AMDGPU_TARGETS[@]} > "${target_lst}"
	fi
	export ROCM_TARGET_LST="${target_lst}"
```

With this code I was able to safely remove addpredict in src_configure from rocBLAS and rocSPARSE (and in 6.1.x release there are no more ROCm packages with addpredict).

Note that some other ROCm libraries (not in Gentoo) still call rocminfo (which requires addpredict) without specific reason:
* https://github.com/ROCm/rocAL/blob/rocm-6.2.0/rocAL-setup.py#L80
* https://github.com/ROCm/rocDecode/blob/rocm-6.2.0/rocDecode-setup.py#L63

But for now I don't see that anybody uses these libraries, so no ebuilds for them.